temporal-sdk/src/main/java/io/temporal/internal/sync/DeterministicRunnerImpl.java

```diff
@@ -290,17 +314,23 @@ public void close() {
       lock.unlock();
       return;
     }
-    // Do not close while runUntilAllBlocked executes.
-    // closeRequested tells it to call close() at the end.
     closeRequested = true;
-    if (inRunUntilAllBlocked) {
+    if (
+    // Do not close immediately if runUntilAllBlocked executes.
+    // closeRequested will tell a control thread to call close() at the end of the event loop.
+    inRunUntilAllBlocked
+        // some other thread or caller is already in the process of closing
+        || closeStarted) {
+
       lock.unlock();
+      // We will not perform the closure in this call and should just wait on the future when
+      // another thread responsible for it will close.
+      closeFuture.join();
```

Comment:
Any concerns of a hang here? Specifically, if `closeFuture.complete` is not reached but `closeStarted = true` is, this code will hang.

Reply:
No. If the thread started shutdown, but didn't reach the end - it's a bug and something unexpected happened. But to make this code LOOK safer, I will wrap in it try-finally.